api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2647,49 +2715,45 @@ public void composeObject(
       }
     }
 
-    if (partsCount == 1) {
+    if (args.sources().size() == 1) {
       ComposeSource src = sources.get(0);
-      if (headerMap == null) {
-        headerMap = new HashMap<>();
-      }
-      if ((src.offset() != null) && (src.length() == null)) {
-        headerMap.put("x-amz-copy-source-range", "bytes=" + src.offset() + "-");
-      }
-
-      if ((src.offset() != null) && (src.length() != null)) {
-        headerMap.put(
-            "x-amz-copy-source-range",
-            "bytes=" + src.offset() + "-" + (src.offset() + src.length() - 1));
+      Multimap<String, String> headers = HashMultimap.create();
+      headers.putAll(args.extraHeaders());
+      headers.putAll(args.headers);
+      if (src.offset() == null && src.length() == null) {
+        return copyObject(
+            CopyObjectArgs.builder()
+                .bucket(args.bucket())
+                .object(args.object())
+                .headers(args.extraHeaders())
+                .sse(args.sse())
+                .srcBucket(src.bucket())
+                .srcObject(src.object())
+                .srcVersionId(src.versionId())
+                .srcSsec(src.ssec())
+                .srcMatchETag(src.matchETag())
+                .srcNotMatchETag(src.notMatchETag())
+                .srcModifiedSince(src.modifiedSince())
+                .srcUnmodifiedSince(src.unmodifiedSince())
+                .build());
       }
-      copyObject(
-          bucketName,
-          objectName,
-          headerMap,
-          sse,
-          src.bucketName(),
-          src.objectName(),
-          src.ssec(),
-          src.copyConditions());
-      return;
     }
 
-    Map<String, String> sseHeaders = null;
-    if (sse != null) {
-      sseHeaders = sse.headers();
-      if (headerMap == null) {
-        headerMap = new HashMap<>();
-      }
-      headerMap.putAll(sseHeaders);
+    Multimap<String, String> sseHeaders = HashMultimap.create();
+    Multimap<String, String> headerMap = HashMultimap.create();
+    if (args.sse() != null) {
+      sseHeaders.putAll(Multimaps.forMap(args.sse().headers()));
+      headerMap.putAll(args.extraHeaders());
+      headerMap.putAll(args.headers);
     }
 
-    String uploadId = createMultipartUpload(bucketName, objectName, headerMap);
+    String uploadId = createMultipartUpload(args.bucket(), null, args.object(), headerMap, null);
```

Comment:
1. use `createMultipartUpload()` like `String uploadId = createMultipartUpload(args.bucket(), args.region(), args.object(), args.genHeaders(), args.extraQueryParams());`
2. move this line at 1242